### PR TITLE
Attempt to retarget tests to .NET 6.0

### PR DESCRIPTION
**Pull request by ericstj**
Trying out retargeting our tests to .NET 6.0.

### CODE REVIEW

**Comment by ericstj** [comment]
I believe we're tripping up on the differences between CPUMath and Intrinsics.  Either we're using the wrong data file, or we're loading the wrong CPUMath.  Looking into it.

**Comment by ericstj** [comment]
I think I found and fixed the problems around test data.  We weren't even using the .NETCore 3.1+ data since the way we were checking for .NETCore 3.1+ broke in 6.0.  I fixed that.  Also fixed a few other cases where we were conditioning our behavior based on .NETCore 3.1.  I made all the chanages here "future proof" by avoiding hardcoding a version we'll have to change moving forward.

**Comment by ericstj** [comment]
This seems to have fixed the majority of the issues, here are the two remaining, across all platforms:
MulticlassLRNonNegativeTest
```
[m[37m        *** Failure #1: Values to compare are 0.1270046062939375 and 0.1273098059294578
[m[37m        	 AllowedVariance: 0.0001
[m[37m        	 delta: -0.0003
[m[37m        	 delta2: -0.0003

[m[37m        *** Failure #2: Output and baseline mismatch at line 43, expected '41	0	0	0.12700460629393751	0.8807296	0.11926952	7.253493E-08	0	1	2' but got '41	0	0	0.12730980592945779	0.88046086	0.119539626	7.181209E-08	0	1	2' : 'MulticlassLogisticRegression/LogisticRegression-Non-Negative-TrainTest-iris.txt'

[m[37m        *** Failure #3: Values to compare are 0.30088424491217947 and 0.3007240049803194
[m[37m        	 AllowedVariance: 0.0001
[m[37m        	 delta: 0.0002
[m[37m        	 delta2: 0.0002

[m[37m        *** Failure #4: Output and baseline mismatch at line 32, expected '68	1	1	0.30088424491217947	0.740163445	0.258734822	0.00110106682	1	2	0' but got '68	1	1	0.30072400498031943	0.74028206	0.25861543	0.001101945	1	2	0' : 'MulticlassLogisticRegression/LogisticRegression-Non-Negative-CV-iris.txt'
```

BinaryClassifierLogisticRegressionBinNormTest 
```
[m[37m        *** Failure #1: Values to compare are -5.641634 and -5.6416364
[m[37m        	 AllowedVariance: 1E-06
[m[37m        	 delta: 1E-05
[m[37m        	 delta2: 2E-06

[m[37m        *** Failure #2: Output and baseline mismatch at line 12, expected '10	0	-5.641634	0.00353453052	0.0051082826744376919	0' but got '10	0	-5.6416364	0.0035345221	0.0051082705390145811	0' : 'LogisticRegression/LogisticRegression-bin-norm-TrainTest-breast-cancer.txt'

[m[37m        *** Failure #3: Values to compare are 6.373105 and 6.3731103
[m[37m        	 AllowedVariance: 1E-06
[m[37m        	 delta: -1E-05
[m[37m        	 delta2: -5E-06

[m[37m        *** Failure #4: Output and baseline mismatch at line 2, expected '5	1	6.373105	0.9982961	0.0024603307167519865	1' but got '5	1	6.3731103	0.9982961	0.0024603307167519865	1' : 'LogisticRegression/LogisticRegression-bin-norm-CV-breast-cancer.txt'
```

These small changes seem like acceptable rounding error to me.  I'm inclined to just update the .NET Core baseline files  @michaelgsharp do you agree?

**Comment by ericstj** [comment]
Baseline data updated to account for those failures.  We also have CPUMath failures on ARM64.  Looks like those are trying to load `CpuMathNative` which we don't expect.

**Comment by ericstj** [comment]
Ubuntu failure is well known system.drawing issue.
```
Microsoft.ML.Tests.ImageTests.TestBackAndForthConversionWithAlphaInterleave [FAIL][31;1m[m[37m      System.ArgumentException : Parameter is not valid.
[m[30;1m      Stack Trace:
[m[37m           at System.Drawing.Bitmap.UnlockBits(BitmapData bitmapdata)
```
Rest of the legs are looking good 👍 

Coverage is broken for some reason.  I'll look into that next.

**Comment by ericstj** [comment]
I thought the bug here was that our PathResolver for loading from the nuget package cache was busted, but actually it's that the runtime.config.dev.json is no longer produced when targeting net6.0.  This is the piece that tells the runtime how to locate the nuget package folder.  https://learn.microsoft.com/en-us/dotnet/core/compatibility/sdk/6.0/runtimeconfigdev-file

We can fix this by setting `GenerateRuntimeConfigDevFile` to restore the old behavior.

**Comment by ericstj** [comment]
Looks very close.  The one failure on ARM seems to be a transient test failure.  I've opened https://github.com/dotnet/machinelearning/issues/6377 to track it and reran that leg.

**Comment by ericstj** [comment]
image
✋🎉🎆🚀

File: test/BaselineOutput/Common/LogisticRegression/netcoreapp31/LogisticRegression-bin-norm-CV-breast-cancer.txt [hunk 1]
```diff
@@ -0,0 +1,700 @@
+Instance	Label	Score	Probability	Log-loss	Assigned
```
Comment (tarekgh) [inline test/BaselineOutput/Common/LogisticRegression/netcoreapp31/LogisticRegression-bin-norm-CV-breast-cancer.txt:1]: do we need to keep h e folder name as netcoreapp31?

File: test/Microsoft.ML.CpuMath.UnitTests/Microsoft.ML.CpuMath.UnitTests.csproj [hunk 2]
```diff
@@ -5,7 +5,7 @@
     <ProjectReference Include="..\Microsoft.ML.TestFramework\Microsoft.ML.TestFramework.csproj" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' != 'netcoreapp3.1'">
+  <ItemGroup Condition="!$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'netcoreapp3.1'))">
```
Comment (tarekgh) [inline test/Microsoft.ML.CpuMath.UnitTests/Microsoft.ML.CpuMath.UnitTests.csproj:8]: should we just check if it is net core? or not netfx?

File: test/Microsoft.ML.PerformanceTests/Harness/Configs.cs [hunk 3]
```diff
@@ -60,9 +62,7 @@ private IToolchain CreateToolchain()
 
         protected static string GetBuildConfigurationName()
         {
-#if NETCOREAPP3_1
-            return "Release-netcoreapp3_1";
-#elif NET461
+#if NETFRAMEWORK
             return "Release-netfx";
 #else
             return "Release";
```
Comment (tarekgh) [inline test/Microsoft.ML.PerformanceTests/Harness/Configs.cs:68]: >Release [](http://example.com/codeflow?start=20&length=7)

You don't think we need to append this with some specific version? I know this work fine but if we'll have more targets in the future can cause issues?

**Review by tarekgh — APPROVED** [review]

File: test/Microsoft.ML.CpuMath.UnitTests/Microsoft.ML.CpuMath.UnitTests.csproj (hunk 2 above)
Comment (ericstj) [reply test/Microsoft.ML.CpuMath.UnitTests/Microsoft.ML.CpuMath.UnitTests.csproj:8]: No, it’s dependent on the product behavior which starts at NETCore 3.1

File: test/Microsoft.ML.PerformanceTests/Harness/Configs.cs (hunk 3 above)
Comment (ericstj) [reply test/Microsoft.ML.PerformanceTests/Harness/Configs.cs:68]: If we need to change in the future we can.   We should try to remove the custom harness entirely.

File: test/BaselineOutput/Common/LogisticRegression/netcoreapp31/LogisticRegression-bin-norm-CV-breast-cancer.txt (hunk 1 above)
Comment (ericstj) [reply test/BaselineOutput/Common/LogisticRegression/netcoreapp31/LogisticRegression-bin-norm-CV-breast-cancer.txt]: I went ahead and renamed this.  I was trying to keep most of the changes to be "future proof" and not lose the context of which version a change was required for, but that's difficult here without forking more and adding a lot of fallback probing (like look for net6.0, netcoreapp3.1, etc).  I think just removing the version is the best thing to do for now.

**Pull request merged**